mountpoint-s3/src/superblock/path.rs

```diff
@@ -0,0 +1,218 @@
+    NotADirectory,
+}
+
+impl ValidKey {
+    /// Create the root key.
+    pub fn root() -> Self {
+        Self {
+            key: String::new(),
+            name_offset: 0,
+        }
+    }
+
+    /// Create a new child key.
+    pub fn new_child(&self, name: ValidName, kind: InodeKind) -> Result<Self, ValidKeyError> {
+        let InodeKind::Directory = self.kind() else {
+            return Err(ValidKeyError::NotADirectory);
+        };
+
+        let mut key = self.as_ref().to_owned();
+        let name_offset = key.len();
+        key.push_str(&name);
+        if kind == InodeKind::Directory && !key.is_empty() {
+            key.push('/');
+        }
+        Ok(Self { name_offset, key })
+    }
+
+    /// The name for this key, i.e. the last path component.
+    ///
+    /// For directories, the name does not include the terminal '/'.
+    pub fn name(&self) -> &str {
+        let len = self.key.len();
+        if len == 0 {
+            return "";
+        }
+        if self.key.as_bytes()[len - 1] == b'/' {
+            &self.key[self.name_offset..(len - 1)]
+        } else {
+            &self.key[self.name_offset..]
+        }
+    }
+
+    /// The kind of [Inode](super::Inode) associated with this key.
+    pub fn kind(&self) -> InodeKind {
+        match self.key.as_bytes().last() {
+            None | Some(b'/') => InodeKind::Directory,
+            _ => InodeKind::File,
+        }
+    }
+}
+
+impl AsRef<str> for ValidKey {
+    fn as_ref(&self) -> &str {
+        &self.key
+    }
+}
+
+/// A valid name for an [Inode](super::Inode).
+#[derive(Debug, Clone, Copy)]
+pub struct ValidName<'a>(&'a str);
+
+impl<'a> ValidName<'a> {
+    /// Parse a string into a [ValidName].
+    pub fn parse_os_str(name: &'a OsStr) -> Result<Self, InodeError> {
+        let name_str = name.to_str().ok_or_else(|| InodeError::InvalidFileName(name.into()))?;
+        Self::parse_str(name_str)
+    }
+
+    /// Parse a string into a [ValidName].
+    pub fn parse_str(name: &'a str) -> Result<Self, InodeError> {
+        // Names cannot be empty
+        if !name.is_empty() &&
+            // "." and ".." are reserved names (presented by the filesystem layer)
+            name != "." &&
+            name != ".." &&
+            // The delimiter / can never appear in a name
+            !name.as_bytes().contains(&b'/') &&
+            // NUL is invalid in POSIX names
+            !name.as_bytes().contains(&b'\0')
+        {
+            Ok(Self(name))
+        } else {
+            Err(InodeError::InvalidFileName(name.into()))
+        }
+    }
+}
+
+impl<'a> TryFrom<&'a OsStr> for ValidName<'a> {
+    type Error = InodeError;
+
+    fn try_from(value: &'a OsStr) -> Result<Self, Self::Error> {
+        Self::parse_os_str(value)
+    }
+}
+
+impl<'a> TryFrom<&'a str> for ValidName<'a> {
+    type Error = InodeError;
+
+    fn try_from(value: &'a str) -> Result<Self, Self::Error> {
+        Self::parse_str(value)
+    }
+}
+
+impl Deref for ValidName<'_> {
```

Comment:
a bit non-intuitive that `*valid_name` and `&valid_name` are both `&str`, where do we need `Deref` implementation?

Reply:
`Deref` also gives us all the methods on `&str`, see [Deref coercion](https://doc.rust-lang.org/std/ops/trait.Deref.html#deref-coercion).